core/components/menu/menuitem.js

```diff
@@ -25,121 +24,108 @@ goog.require('Blockly.utils.object');
  *     the item.
  * @param {string=} opt_value Data/model associated with the menu item.
  * @constructor
- * @extends {Blockly.Component}
  */
 Blockly.MenuItem = function(content, opt_value) {
-  Blockly.Component.call(this);
+  /**
+   * Human-readable text of this menu item.
+   * @type {string}
+   * @private
+   */
+  this.content_ = content;
 
-  this.setContentInternal(content);
-  this.setValue(opt_value);
+  /**
+   * Machine-readable value of this menu item.
+   * @type {string|undefined}
+   * @private
+   */
+  this.value_ = opt_value;
 
   /**
+   * Is the menu item clickable, as opposed to greyed-out.
    * @type {boolean}
    * @private
    */
   this.enabled_ = true;
 
   /**
-   * @type {Blockly.MenuItem}
+   * The DOM element for the menu item.
+   * @type {?Element}
    * @private
    */
-  this.previousSibling_;
+  this.element_ = null;
 
   /**
-   * @type {Blockly.MenuItem}
-   * @private
+   * Whether the menu item is rendered right-to-left.
+   * @type {boolean}
+   * @protected
    */
-  this.nextSibling_;
+  this.rightToLeft_ = false;
 };
-Blockly.utils.object.inherits(Blockly.MenuItem, Blockly.Component);
 
 
 /**
  * Creates the menuitem's DOM.
- * @override
+ * @return {!Element} Completed DOM.
  */
 Blockly.MenuItem.prototype.createDom = function() {
   var element = document.createElement('div');
-  element.id = this.getId();
-  this.setElementInternal(element);
+  element.id = Blockly.utils.IdGenerator.getNextUniqueId();
+  element.blocklyMenuItem = this;  // Link DOM back to this data structure.
```

Comment:
Frankly, I'm not a huge fan of this pattern, as it shortcuts typing. There is precedence for this pattern in Blockly (eg: tooltips), so not saying it has to be changed. But here are my thoughts on the pattern:
- It makes it harder to debug for a someone new to the team / area, as there's hidden context (traditionally a DOM element doesn't have additional properties that are not DOM properties)
- The closure typechecker is not complaining (currently) because strictTypeChecks are not on, but when we do eventually get there, this will introduce issues. There are ways around that, ie: Create a new type that subclasses HTMLElement, and adds a blocklyMenuItem of type MenuItem as a property, but its clunky. 

That being said, if the perf benefit is worth it for you, then by all means keep as is.

Reply:
The original Closure code maintained a database for efficient look-ups.  But it was a lot of code and complexity undeserving for this task.  The DOM-to-JS links were simple and efficient, but have the problems you cite.  I've replaced this with an O(n) search of the menu's options.  Shouldn't be a problem given that our menus/dropdowns aren't intended to have hundreds or thousands of items.